File: clients/ios/Lextures/Core/LMS/CourseGradingLogic.swift
import Foundation

/// Course grading scale, weighted groups, and item mapping helpers (M13.4).
enum CourseGradingLogic {
    struct GradingScaleOption: Identifiable, Hashable {
        var id: String
        var labelKey: String
        var descriptionKey: String
    }

    struct SchemeDisplayType: Identifiable, Hashable {
        var id: String
        var labelKey: String
    }

    struct EditableAssignmentGroup: Equatable, Hashable {
        var clientKey: String
        var id: String?
        var name: String
        var sortOrder: Int
        var weightPercent: String
    }

    struct GradingSchemeBand: Identifiable, Equatable, Hashable {
        var clientKey: String
        var label: String
        var minPct: String
        var gpa: String

        var id: String { clientKey }
    }

    struct GradableRow: Identifiable, Hashable {
        var item: CourseStructureItem
        var moduleTitle: String
        var id: String { item.id }
    }

    struct FormBaseline: Equatable {
        var gradingScale: String
        var groups: [EditableAssignmentGroup]
        var schemeType: String
        var bands: [GradingSchemeBand]
        var passMinPct: String
        var completeMinPct: String
    }

    enum ValidationError: Equatable {
        case groupsNeedNames
        case bandsInvalid(String)
        case schemeInvalid(String)
    }

    static let gradingScaleOptions: [GradingScaleOption] = [
        .init(id: "letter_standard", labelKey: "mobile.courseSettings.grading.scale.letterStandard.label", descriptionKey: "mobile.courseSettings.grading.scale.letterStandard.description"),
        .init(id: "letter_plus_minus", labelKey: "mobile.courseSettings.grading.scale.letterPlusMinus.label", descriptionKey: "mobile.courseSettings.grading.scale.letterPlusMinus.description"),
        .init(id: "percent", labelKey: "mobile.courseSettings.grading.scale.percent.label", descriptionKey: "mobile.courseSettings.grading.scale.percent.description"),
        .init(id: "pass_fail", labelKey: "mobile.courseSettings.grading.scale.passFail.label", descriptionKey: "mobile.courseSettings.grading.scale.passFail.description"),
    ]

    static let schemeDisplayTypes: [SchemeDisplayType] = [
        .init(id: "points", labelKey: "mobile.courseSettings.grading.scheme.type.points"),
        .init(id: "percentage", labelKey: "mobile.courseSettings.grading.scheme.type.percentage"),
        .init(id: "letter", labelKey: "mobile.courseSettings.grading.scheme.type.letter"),
        .init(id: "gpa", labelKey: "mobile.courseSettings.grading.scheme.type.gpa"),
        .init(id: "pass_fail", labelKey: "mobile.courseSettings.grading.scheme.type.passFail"),
        .init(id: "complete_incomplete", labelKey: "mobile.courseSettings.grading.scheme.type.completeIncomplete"),
    ]

    static func cacheKeyGrading(courseCode: String) -> String {
        "course:\(courseCode):grading-settings"
    }

    static func settingsIdempotencyKey(courseCode: String) -> String {
        "course-grading:\(courseCode):settings"
    }

    static func schemeIdempotencyKey(courseCode: String) -> String {
        "course-grading:\(courseCode):scheme"
    }

    static func itemMappingIdempotencyKey(courseCode: String, itemId: String) -> String {
        "course-grading:\(courseCode):item-group:\(itemId)"
    }

    static func newClientKey() -> String {
        "new-\(UUID().uuidString)"
    }

    static func defaultGroups() -> [EditableAssignmentGroup] {
        [
            .init(
                clientKey: newClientKey(),
                id: nil,
                name: "Assignments",
                sortOrder: 0,
                weightPercent: "100"
            ),
        ]
    }

    static func defaultBands() -> [GradingSchemeBand] {
        [
            .init(clientKey: newClientKey(), label: "A", minPct: "90", gpa: "4"),
            .init(clientKey: newClientKey(), label: "B", minPct: "80", gpa: "3"),
            .init(clientKey: newClientKey(), label: "C", minPct: "70", gpa: "2"),
            .init(clientKey: newClientKey(), label: "D", minPct: "60", gpa: "1"),
            .init(clientKey: newClientKey(), label: "F", minPct: "0", gpa: "0"),
        ]
    }

    static func groupsFromSettings(_ settings: CourseGradingSettings) -> [EditableAssignmentGroup] {
        if settings.assignmentGroups.isEmpty { return defaultGroups() }
        return settings.assignmentGroups.map { group in
            EditableAssignmentGroup(
                clientKey: group.id,
                id: group.id,
                name: group.name,
                sortOrder: group.sortOrder,
                weightPercent: String(group.weightPercent)
            )
        }
    }

    static func baseline(
        settings: CourseGradingSettings,
        scheme: CourseGradingSchemeRecord?
    ) -> FormBaseline {
        let schemeType = scheme?.type.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty ?? "points"
        let parsed = parseBands(from: scheme?.scaleJson)
        let passMin = parsePassMinPct(from: scheme?.scaleJson) ?? "60"
        let completeMin = parseCompleteMinPct(from: scheme?.scaleJson) ?? "50"
        return FormBaseline(
            gradingScale: settings.gradingScale.nilIfEmpty ?? "letter_standard",
            groups: groupsFromSettings(settings),
            schemeType: schemeType,
            bands: parsed.isEmpty ? defaultBands() : parsed,
            passMinPct: passMin,
            completeMinPct: completeMin
        )
    }

    static func weightTotal(_ groups: [EditableAssignmentGroup]) -> Double {
        var total = 0.0
        for group in groups {
            let value = Double(group.weightPercent.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
            if value.isFinite { total += value }
        }
        return (total * 1000).rounded() / 1000
    }

    static func hasWeightWarning(_ total: Double) -> Bool {
        abs(total - 100) >= 0.01
    }

    static func weightTotalLabel(total: Double) -> String {
        L.format("mobile.courseSettings.grading.weightTotal", String(format: "%.2f", total))
    }

    static func gradableRows(from structure: [CourseStructureItem]) -> [GradableRow] {
        var rows: [GradableRow] = []
        var moduleTitle = ""
        for item in structure {
            if item.kind == "module" {
                moduleTitle = item.title
            } else if item.isGradable {
                rows.append(.init(item: item, moduleTitle: moduleTitle))
            }
        }
        return rows
    }

    static func namedGroupsWithIds(_ groups: [EditableAssignmentGroup]) -> [EditableAssignmentGroup] {
        groups.filter { !$0.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty && $0.id != nil }
    }

    static func isSettingsDirty(current: FormBaseline, baseline: FormBaseline) -> Bool {
        if current.gradingScale != baseline.gradingScale { return true }
        return normalizedGroups(current.groups) != normalizedGroups(baseline.groups)
    }

    static func isSchemeDirty(current: FormBaseline, baseline: FormBaseline) -> Bool {
        if current.schemeType != baseline.schemeType { return true }
        if current.schemeType == "letter" || current.schemeType == "gpa" {
            return normalizedBands(current.bands) != normalizedBands(baseline.bands)
        }
        if current.schemeType == "pass_fail" {
            return current.passMinPct.trimmingCharacters(in: .whitespacesAndNewlines)
                != baseline.passMinPct.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if current.schemeType == "complete_incomplete" {
            return current.completeMinPct.trimmingCharacters(in: .whitespacesAndNewlines)
                != baseline.completeMinPct.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return false
    }

    static func validateGroups(_ groups: [EditableAssignmentGroup]) -> ValidationError? {
        let named = groups.filter { !$0.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        if named.isEmpty || named.count != groups.count {
            return .groupsNeedNames
        }
        return nil
    }

    static func validateBands(_ bands: [GradingSchemeBand]) -> ValidationError? {
        guard !bands.isEmpty else {
            return .bandsInvalid(L.text("mobile.courseSettings.grading.validation.bandsRequired"))
        }
        var parsed: [(label: String, minPct: Double)] = []
        for (index, band) in bands.enumerated() {
            let label = band.label.trimmingCharacters(in: .whitespacesAndNewlines)
            if label.isEmpty {
                return .bandsInvalid(L.format("mobile.courseSettings.grading.validation.bandLabelRequired", index + 1))
            }
            guard let minPct = Double(band.minPct.trimmingCharacters(in: .whitespacesAndNewlines)),
                  minPct.isFinite, minPct >= 0, minPct <= 100 else {
                return .bandsInvalid(L.format("mobile.courseSettings.grading.validation.bandMinOutOfRange", index + 1))
            }
            parsed.append((label, minPct))
        }
        let ascending = parsed.sorted { $0.minPct < $1.minPct }
        if abs(ascending[0].minPct) > 0.001 {
            return .bandsInvalid(L.text("mobile.courseSettings.grading.validation.lowestBandMustBeZero"))
        }
        for index in 1 ..< ascending.count where ascending[index].minPct <= ascending[index - 1].minPct + 0.001 {
            return .bandsInvalid(L.text("mobile.courseSettings.grading.validation.bandsMustIncrease"))
        }
        return nil
    }

    static func validateScheme(form: FormBaseline) -> ValidationError? {
        switch form.schemeType {
        case "letter", "gpa":
            return validateBands(form.bands)
        case "pass_fail":
            guard let value = Double(form.passMinPct.trimmingCharacters(in: .whitespacesAndNewlines)),
                  value.isFinite, value >= 0, value <= 100 else {
                return .schemeInvalid(L.text("mobile.courseSettings.grading.validation.passMinOutOfRange"))
            }
            return nil
        case "complete_incomplete":
            guard let value = Double(form.completeMinPct.trimmingCharacters(in: .whitespacesAndNewlines)),
                  value.isFinite, value >= 0, value <= 100 else {
                return .schemeInvalid(L.text("mobile.courseSettings.grading.validation.completeMinOutOfRange"))
            }
            return nil
        default:
            return nil
        }
    }

    static func buildPutSettingsBody(form: FormBaseline) -> PutCourseGradingSettingsBody {
        PutCourseGradingSettingsBody(
            gradingScale: form.gradingScale,
            assignmentGroups: form.groups.enumerated().map { index, group in
                let weight = Double(group.weightPercent.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
                return CourseAssignmentGroupInput(
                    id: group.id,
                    name: group.name.trimmingCharacters(in: .whitespacesAndNewlines),
                    sortOrder: index,
                    weightPercent: weight.isFinite ? weight : 0,
                    dropLowest: 0,
                    dropHighest: 0,
                    replaceLowestWithFinal: false
                )
            }
        )
    }

    static func buildPutSchemeBody(form: FormBaseline) -> PutCourseGradingSchemeBody {
        let scaleJson: JSONValue?
        switch form.schemeType {
        case "letter", "gpa":
            scaleJson = encodeBands(form.bands)
        case "pass_fail":
            let value = Double(form.passMinPct.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 60
            scaleJson = .object(["pass_min_pct": .number(value)])
        case "complete_incomplete":
            let value = Double(form.completeMinPct.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 50
            scaleJson = .object(["complete_min_pct": .number(value)])
        default:
            scaleJson = .object([:])
        }
        return PutCourseGradingSchemeBody(type: form.schemeType, scaleJson: scaleJson)
    }

    static func sortBandsDescending(_ bands: [GradingSchemeBand]) -> [GradingSchemeBand] {
        bands.sorted {
            let left = Double($0.minPct.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
            let right = Double($1.minPct.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
            return left > right
        }
    }

    static func kindLabelKey(for kind: String) -> String {
        switch kind {
        case "quiz": return "mobile.courseSettings.grading.mapping.type.quiz"
        case "content_page": return "mobile.courseSettings.grading.mapping.type.content"
        default: return "mobile.courseSettings.grading.mapping.type.assignment"
        }
    }

    private static func normalizedGroups(_ groups: [EditableAssignmentGroup]) -> [[String: String]] {
        groups.map { group in
            [
                "name": group.name.trimmingCharacters(in: .whitespacesAndNewlines),
                "weight": String(Double(group.weightPercent.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0),
            ]
        }
    }

    private static func normalizedBands(_ bands: [GradingSchemeBand]) -> [[String: String]] {
        sortBandsDescending(bands).map { band in
            [
                "label": band.label.trimmingCharacters(in: .whitespacesAndNewlines),
                "min_pct": band.minPct.trimmingCharacters(in: .whitespacesAndNewlines),
                "gpa": band.gpa.trimmingCharacters(in: .whitespacesAndNewlines),
            ]
        }
    }

    private static func parseBands(from scaleJson: JSONValue?) -> [GradingSchemeBand] {
        guard case let .array(items) = scaleJson else { return [] }
        return items.compactMap { value -> GradingSchemeBand? in
            guard case let .object(obj) = value else { return nil }
            let label: String
            if case let .string(text) = obj["label"] ?? .null { label = text } else { return nil }
            let minPct: String
            if case let .number(number) = obj["min_pct"] ?? .null {
                minPct = String(number)
            } else if case let .string(text) = obj["min_pct"] ?? .null {
                minPct = text
            } else {
                return nil
            }
            let gpa: String
            if case let .number(number) = obj["gpa"] ?? .null {
                gpa = String(number)
            } else if case let .string(text) = obj["gpa"] ?? .null {
                gpa = text
            } else {
                gpa = ""
            }
            return GradingSchemeBand(clientKey: newClientKey(), label: label, minPct: minPct, gpa: gpa)
        }
    }

    private static func parsePassMinPct(from scaleJson: JSONValue?) -> String? {
        guard case let .object(obj) = scaleJson else { return nil }
        if case let .number(number) = obj["pass_min_pct"] ?? .null { return String(number) }
        if case let .string(text) = obj["pass_min_pct"] ?? .null { return text }
        return nil
    }

    private static func parseCompleteMinPct(from scaleJson: JSONValue?) -> String? {
        guard case let .object(obj) = scaleJson else { return nil }
        if case let .number(number) = obj["complete_min_pct"] ?? .null { return String(number) }
        if case let .string(text) = obj["complete_min_pct"] ?? .null { return text }
        return nil
    }

    private static func encodeBands(_ bands: [GradingSchemeBand]) -> JSONValue {
        .array(
            sortBandsDescending(bands).map { band in
                var object: [String: JSONValue] = [
                    "label": .string(band.label.trimmingCharacters(in: .whitespacesAndNewlines)),
                    "min_pct": .number(Double(band.minPct.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0),
                ]
                if let gpa = Double(band.gpa.trimmingCharacters(in: .whitespacesAndNewlines)) {
                    object["gpa"] = .number(gpa)
                }
                return .object(object)
            }
        )
    }
}

private extension String {
    var nilIfEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CourseGradingLogic.kt
package com.lextures.android.core.lms

import androidx.annotation.StringRes
import com.lextures.android.R
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.util.UUID
import kotlin.math.abs

/** Course grading scale, weighted groups, and item mapping helpers (M13.4). */
object CourseGradingLogic {
    enum class GradingScaleOptionId {
        letter_standard,
        letter_plus_minus,
        percent,
        pass_fail,
    }

    enum class SchemeDisplayTypeId {
        points,
        percentage,
        letter,
        gpa,
        pass_fail,
        complete_incomplete,
    }

    data class GradingScaleOption(val id: GradingScaleOptionId)
    data class SchemeDisplayType(val id: SchemeDisplayTypeId)

    data class EditableAssignmentGroup(
        val clientKey: String,
        val id: String? = null,
        val name: String,
        val sortOrder: Int,
        val weightPercent: String,
    )

    data class GradingSchemeBand(
        val clientKey: String,
        val label: String,
        val minPct: String,
        val gpa: String = "",
    )

    data class GradableRow(val item: CourseStructureItem, val moduleTitle: String)

    data class FormBaseline(
        val gradingScale: String,
        val groups: List<EditableAssignmentGroup>,
        val schemeType: String,
        val bands: List<GradingSchemeBand>,
        val passMinPct: String,
        val completeMinPct: String,
    )

    sealed class ValidationError {
        data object GroupsNeedNames : ValidationError()
        data class BandsInvalid(val message: String) : ValidationError()
        data class SchemeInvalid(val message: String) : ValidationError()
    }

    val gradingScaleOptions = listOf(
        GradingScaleOption(GradingScaleOptionId.letter_standard),
        GradingScaleOption(GradingScaleOptionId.letter_plus_minus),
        GradingScaleOption(GradingScaleOptionId.percent),
        GradingScaleOption(GradingScaleOptionId.pass_fail),
    )

    val schemeDisplayTypes = listOf(
        SchemeDisplayType(SchemeDisplayTypeId.points),
        SchemeDisplayType(SchemeDisplayTypeId.percentage),
        SchemeDisplayType(SchemeDisplayTypeId.letter),
        SchemeDisplayType(SchemeDisplayTypeId.gpa),
        SchemeDisplayType(SchemeDisplayTypeId.pass_fail),
        SchemeDisplayType(SchemeDisplayTypeId.complete_incomplete),
    )

    @StringRes
    fun gradingScaleLabelRes(id: GradingScaleOptionId): Int = when (id) {
        GradingScaleOptionId.letter_standard -> R.string.mobile_courseSettings_grading_scale_letterStandard_label
        GradingScaleOptionId.letter_plus_minus -> R.string.mobile_courseSettings_grading_scale_letterPlusMinus_label
        GradingScaleOptionId.percent -> R.string.mobile_courseSettings_grading_scale_percent_label
        GradingScaleOptionId.pass_fail -> R.string.mobile_courseSettings_grading_scale_passFail_label
    }

    @StringRes
    fun gradingScaleDescriptionRes(id: GradingScaleOptionId): Int = when (id) {
        GradingScaleOptionId.letter_standard -> R.string.mobile_courseSettings_grading_scale_letterStandard_description
        GradingScaleOptionId.letter_plus_minus -> R.string.mobile_courseSettings_grading_scale_letterPlusMinus_description
        GradingScaleOptionId.percent -> R.string.mobile_courseSettings_grading_scale_percent_description
        GradingScaleOptionId.pass_fail -> R.string.mobile_courseSettings_grading_scale_passFail_description
    }

    @StringRes
    fun schemeTypeLabelRes(id: SchemeDisplayTypeId): Int = when (id) {
        SchemeDisplayTypeId.points -> R.string.mobile_courseSettings_grading_scheme_type_points
        SchemeDisplayTypeId.percentage -> R.string.mobile_courseSettings_grading_scheme_type_percentage
        SchemeDisplayTypeId.letter -> R.string.mobile_courseSettings_grading_scheme_type_letter
        SchemeDisplayTypeId.gpa -> R.string.mobile_courseSettings_grading_scheme_type_gpa
        SchemeDisplayTypeId.pass_fail -> R.string.mobile_courseSettings_grading_scheme_type_passFail
        SchemeDisplayTypeId.complete_incomplete -> R.string.mobile_courseSettings_grading_scheme_type_completeIncomplete
    }

    @StringRes
    fun kindLabelRes(kind: String): Int = when (kind) {
        "quiz" -> R.string.mobile_courseSettings_grading_mapping_type_quiz
        "content_page" -> R.string.mobile_courseSettings_grading_mapping_type_content
        else -> R.string.mobile_courseSettings_grading_mapping_type_assignment
    }

    fun cacheKeyGrading(courseCode: String): String = "course:$courseCode:grading-settings"
    fun settingsIdempotencyKey(courseCode: String): String = "course-grading:$courseCode:settings"
    fun schemeIdempotencyKey(courseCode: String): String = "course-grading:$courseCode:scheme"
    fun itemMappingIdempotencyKey(courseCode: String, itemId: String): String = "course-grading:$courseCode:item-group:$itemId"

    fun newClientKey(): String = "new-${UUID.randomUUID()}"

    fun defaultGroups(): List<EditableAssignmentGroup> = listOf(
        EditableAssignmentGroup(newClientKey(), null, "Assignments", 0, "100"),
    )

    fun defaultBands(): List<GradingSchemeBand> = listOf(
        GradingSchemeBand(newClientKey(), "A", "90", "4"),
        GradingSchemeBand(newClientKey(), "B", "80", "3"),
        GradingSchemeBand(newClientKey(), "C", "70", "2"),
        GradingSchemeBand(newClientKey(), "D", "60", "1"),
        GradingSchemeBand(newClientKey(), "F", "0", "0"),
    )

    fun groupsFromSettings(settings: CourseGradingSettings): List<EditableAssignmentGroup> {
        if (settings.assignmentGroups.isEmpty()) return defaultGroups()
        return settings.assignmentGroups.map { group ->
            EditableAssignmentGroup(group.id, group.id, group.name, group.sortOrder, group.weightPercent.toString())
        }
    }

    fun baseline(settings: CourseGradingSettings, scheme: CourseGradingSchemeRecord?): FormBaseline {
        val schemeType = scheme?.type?.trim()?.takeIf { it.isNotEmpty() } ?: "points"
        val parsed = parseBands(scheme?.scaleJson)
        return FormBaseline(
            gradingScale = settings.gradingScale.trim().ifEmpty { "letter_standard" },
            groups = groupsFromSettings(settings),
            schemeType = schemeType,
            bands = parsed.ifEmpty { defaultBands() },
            passMinPct = parsePassMinPct(scheme?.scaleJson) ?: "60",
            completeMinPct = parseCompleteMinPct(scheme?.scaleJson) ?: "50",
        )
    }

    fun weightTotal(groups: List<EditableAssignmentGroup>): Double {
        val total = groups.sumOf { group ->
            group.weightPercent.trim().toDoubleOrNull()?.takeIf { it.isFinite() } ?: 0.0
        }
        return kotlin.math.round(total * 1000.0) / 1000.0
    }

    fun hasWeightWarning(total: Double): Boolean = abs(total - 100.0) >= 0.01

    fun gradableRows(structure: List<CourseStructureItem>): List<GradableRow> {
        val rows = mutableListOf<GradableRow>()
        var moduleTitle = ""
        structure.forEach { item ->
            if (item.kind == "module") moduleTitle = item.title
            else if (item.isGradable) rows += GradableRow(item, moduleTitle)
        }
        return rows
    }

    fun namedGroupsWithIds(groups: List<EditableAssignmentGroup>): List<EditableAssignmentGroup> =
        groups.filter { it.name.trim().isNotEmpty() && it.id != null }

    fun isSettingsDirty(current: FormBaseline, baseline: FormBaseline): Boolean {
        if (current.gradingScale != baseline.gradingScale) return true
        return normalizedGroups(current.groups) != normalizedGroups(baseline.groups)
    }

    fun isSchemeDirty(current: FormBaseline, baseline: FormBaseline): Boolean {
        if (current.schemeType != baseline.schemeType) return true
        return when (current.schemeType) {
            "letter", "gpa" -> normalizedBands(current.bands) != normalizedBands(baseline.bands)
            "pass_fail" -> current.passMinPct.trim() != baseline.passMinPct.trim()
            "complete_incomplete" -> current.completeMinPct.trim() != baseline.completeMinPct.trim()
            else -> false
        }
    }

    fun validateGroups(groups: List<EditableAssignmentGroup>): ValidationError? {
        val named = groups.filter { it.name.trim().isNotEmpty() }
        return if (named.isEmpty() || named.size != groups.size) ValidationError.GroupsNeedNames else null
    }

    fun validateBands(bands: List<GradingSchemeBand>): ValidationError? {
        if (bands.isEmpty()) return ValidationError.BandsInvalid("bands-required")
        val parsed = mutableListOf<Pair<String, Double>>()
        bands.forEachIndexed { index, band ->
            val label = band.label.trim()
            if (label.isEmpty()) return ValidationError.BandsInvalid("band-label-$index")
            val minPct = band.minPct.trim().toDoubleOrNull()
            if (minPct == null || !minPct.isFinite() || minPct < 0 || minPct > 100) {
                return ValidationError.BandsInvalid("band-min-$index")
            }
            parsed += label to minPct
        }
        val ascending = parsed.sortedBy { it.second }
        if (abs(ascending.first().second) > 0.001) return ValidationError.BandsInvalid("lowest-zero")
        for (index in 1 until ascending.size) {
            if (ascending[index].second <= ascending[index - 1].second + 0.001) {
                return ValidationError.BandsInvalid("bands-increase")
            }
        }
        return null
    }

    fun validateScheme(form: FormBaseline): ValidationError? = when (form.schemeType) {
        "letter", "gpa" -> validateBands(form.bands)
        "pass_fail" -> {
            val value = form.passMinPct.trim().toDoubleOrNull()
            if (value == null || !value.isFinite() || value < 0 || value > 100) ValidationError.SchemeInvalid("pass-min")
            else null
        }
        "complete_incomplete" -> {
            val value = form.completeMinPct.trim().toDoubleOrNull()
            if (value == null || !value.isFinite() || value < 0 || value > 100) ValidationError.SchemeInvalid("complete-min")
            else null
        }
        else -> null
    }

    fun buildPutSettingsBody(form: FormBaseline): PutCourseGradingSettingsBody =
        PutCourseGradingSettingsBody(
            gradingScale = form.gradingScale,
            assignmentGroups = form.groups.mapIndexed { index, group ->
                val weight = group.weightPercent.trim().toDoubleOrNull()?.takeIf { it.isFinite() } ?: 0.0
                CourseAssignmentGroupInput(
                    id = group.id,
                    name = group.name.trim(),
                    sortOrder = index,
                    weightPercent = weight,
                )
            },
        )

    fun buildPutSchemeBody(form: FormBaseline): PutCourseGradingSchemeBody =
        PutCourseGradingSchemeBody(
            type = form.schemeType,
            scaleJson = when (form.schemeType) {
                "letter", "gpa" -> encodeBands(form.bands)
                "pass_fail" -> buildJsonObject { put("pass_min_pct", JsonPrimitive(form.passMinPct.trim().toDoubleOrNull() ?: 60.0)) }
                "complete_incomplete" -> buildJsonObject { put("complete_min_pct", JsonPrimitive(form.completeMinPct.trim().toDoubleOrNull() ?: 50.0)) }
                else -> buildJsonObject { }
            },
        )

    fun sortBandsDescending(bands: List<GradingSchemeBand>): List<GradingSchemeBand> =
        bands.sortedByDescending { it.minPct.trim().toDoubleOrNull() ?: 0.0 }

    private fun normalizedGroups(groups: List<EditableAssignmentGroup>): List<Pair<String, String>> =
        groups.map { it.name.trim() to (it.weightPercent.trim().toDoubleOrNull() ?: 0.0).toString() }

    private fun normalizedBands(bands: List<GradingSchemeBand>): List<Triple<String, String, String>> =
        sortBandsDescending(bands).map { Triple(it.label.trim(), it.minPct.trim(), it.gpa.trim()) }

    private fun parseBands(scaleJson: JsonElement?): List<GradingSchemeBand> {
        val array = scaleJson as? JsonArray ?: return emptyList()
        return array.mapNotNull { element ->
            val obj = element.jsonObject
            val label = obj["label"]?.jsonPrimitive?.content?.trim().orEmpty()
            val minPct = obj["min_pct"]?.jsonPrimitive?.doubleOrNull?.toString()
                ?: obj["min_pct"]?.jsonPrimitive?.content
            if (label.isEmpty() || minPct.isNullOrBlank()) return@mapNotNull null
            val gpa = obj["gpa"]?.jsonPrimitive?.doubleOrNull?.toString()
                ?: obj["gpa"]?.jsonPrimitive?.content.orEmpty()
            GradingSchemeBand(newClientKey(), label, minPct, gpa)
        }
    }

    private fun parsePassMinPct(scaleJson: JsonElement?): String? =
        (scaleJson as? JsonObject)?.get("pass_min_pct")?.jsonPrimitive?.doubleOrNull?.toString()
            ?: (scaleJson as? JsonObject)?.get("pass_min_pct")?.jsonPrimitive?.content

    private fun parseCompleteMinPct(scaleJson: JsonElement?): String? =
        (scaleJson as? JsonObject)?.get("complete_min_pct")?.jsonPrimitive?.doubleOrNull?.toString()
            ?: (scaleJson as? JsonObject)?.get("complete_min_pct")?.jsonPrimitive?.content

    private fun encodeBands(bands: List<GradingSchemeBand>): JsonArray = buildJsonArray {
        sortBandsDescending(bands).forEach { band ->
            add(
                buildJsonObject {
                    put("label", JsonPrimitive(band.label.trim()))
                    put("min_pct", JsonPrimitive(band.minPct.trim().toDoubleOrNull() ?: 0.0))
                    band.gpa.trim().toDoubleOrNull()?.let { put("gpa", JsonPrimitive(it)) }
                },
            )
        }
    }
}